File: src/test/java/com/ploy/bubble_server_v3/common/exception/CustomAuthenticationEntryPointTest.java
package com.ploy.bubble_server_v3.common.exception;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;

import java.io.IOException;
import java.io.PrintWriter;

import static org.mockito.Mockito.*;
import static org.junit.jupiter.api.Assertions.*;

class CustomAuthenticationEntryPointTest {

    private HttpServletRequest request;
    private HttpServletResponse response;
    private PrintWriter writer;
    private CustomAuthenticationEntryPoint entryPoint;
    private ObjectMapper objectMapper;

    @BeforeEach
    void setUp() {
        // given
        request = mock(HttpServletRequest.class);
        response = mock(HttpServletResponse.class);
        writer = mock(PrintWriter.class);

        entryPoint = new CustomAuthenticationEntryPoint();

        objectMapper = new ObjectMapper();
        objectMapper.registerModule(new JavaTimeModule());
    }

    @Test
    void commence_인증_되지_않은_사용자_응답_반환() throws IOException {
        // given
        when(response.getWriter()).thenReturn(writer);

        // when
        entryPoint.commence(request, response, null);

        // then
        verify(response).setCharacterEncoding("UTF-8");
        verify(response).setContentType("application/json");
        verify(response).setStatus(HttpServletResponse.SC_UNAUTHORIZED);

        ArgumentCaptor<String> responseBodyCaptor = ArgumentCaptor.forClass(String.class);
        verify(writer).write(responseBodyCaptor.capture());

        String responseBody = responseBodyCaptor.getValue();
        assertNotNull(responseBody);

        assertTrue(responseBody.contains("AUTHENTICATION_FAILED"));
        assertTrue(responseBody.contains("사용자 인증에 실패하였습니다."));
    }
}

File: src/test/kotlin/com/ploy/bubble_server_v2/common/exception/CustomAuthenticationEntryPointTest.kt
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.ArgumentCaptor
import org.mockito.Mockito.*
import org.springframework.security.authentication.BadCredentialsException
import java.io.PrintWriter
import com.ploy.bubble_server_v2.common.exception.CustomAuthenticationEntryPoint


class CustomAuthenticationEntryPointTest {

    private lateinit var entryPoint: CustomAuthenticationEntryPoint
    private lateinit var request: HttpServletRequest
    private lateinit var response: HttpServletResponse
    private lateinit var writer: PrintWriter

    @BeforeEach
    fun setup() {
        entryPoint = CustomAuthenticationEntryPoint()
        request = mock(HttpServletRequest::class.java)
        response = mock(HttpServletResponse::class.java)

        // PrintWriter를 Mock 객체로 생성
        writer = mock(PrintWriter::class.java)
        `when`(response.writer).thenReturn(writer)
    }

    @Test
    fun `test commence when authentication fails`() {
        val authException = BadCredentialsException("Authentication Failed")

        `when`(request.requestURI).thenReturn("/test-url")

        entryPoint.commence(request, response, authException)

        // Verify response properties
        verify(response).characterEncoding = "UTF-8"
        verify(response).contentType = "application/json"
        verify(response).status = HttpServletResponse.SC_UNAUTHORIZED

        // Capture the written string
        val writtenString = captureWrittenString()

        // Verify the content of the written string
        assert(writtenString.contains("사용자 인증에 실패하였습니다.")) { "Expected message not found!" }
        assert(writtenString.startsWith("{") && writtenString.endsWith("}")) { "Expected JSON format not found!" }

        // Verify writer interactions
        verify(writer).flush()
        verify(writer).close()
    }

    private fun captureWrittenString(): String {
        // Capture the arguments used in the write method for further inspection
        val captureArgument = ArgumentCaptor.forClass(String::class.java)
        verify(writer).write(captureArgument.capture(), eq(0), anyInt())
        return captureArgument.value
    }
}
